Configuration management for Concourse

blarghmatey opened the pull request:
This PR adds the initial framework for how we will be using PyInfra as well as the implementation of getting a Concourse system installed.

blarghmatey committed:
Initial pass of implementing deployment logic for Concourse

blarghmatey committed:
Adding some basic information about usage of PyInfra for this repository

blarghmatey committed:
Updated concourse config models and added helper for env_var mapping

blarghmatey committed:
Adding pre-commit hook and linting fixes

blarghmatey committed:
Cleaned up and polished concourse config model and added more steps

- Verified functionality of methods and configs in Concourse config models
- Added more steps to deploy logic for Concourse

blarghmatey committed:
Updating README with details of directory layouts.

blarghmatey committed:
Cleaned up Concourse models and started adding concrete deploy

- Cleaned up linting
- Updated Concourse model definitions with some additional parameters
- Added concrete deployment implementation of Concourse installation to validate buil logic

blarghmatey committed:
Added checking for systemd and minor cleanup

shaidar commented on src/ol_configuration_management/components/baseline/setup.py:
```diff
@@ -0,0 +1,15 @@
+        packages=[
+            "curl",
+        ],
+        update=True,
```
Would it be helpful to add `cache_time=3600` or some other value to reduce repetitive requests?

shaidar commented on src/ol_configuration_management/facts/has_systemd.py:
```diff
@@ -0,0 +1,8 @@
+from pyinfra.api import FactBase
```
Maybe adding a comment here about why this file is needed

shaidar commented on src/ol_configuration_management/components/concourse/build.py:
```diff
@@ -0,0 +1,152 @@
+    )
+    # Download latest Concourse release from GitHub
+    concourse_archive = f"https://github.com/concourse/concourse/releases/download/v{concourse_config.version}/concourse-{concourse_config.version}-linux-amd64.tgz"
+    concourse_archive_hash = f"https://github.com/concourse/concourse/releases/download/v{concourse_config.version}/concourse-{concourse_config.version}-linux-amd64.tgz.sha1"
```
Should the URLs live in the model?

blarghmatey committed:
Finished adding config file and key management

- Added handling of populating or generating web and worker key files
- Updated systemd unit
- Added validation and generation of encryption key config

shaidar commented on src/ol_configuration_management/deployments/concourse.py:
```diff
@@ -0,0 +1,25 @@
+
+concourse_config = ConcourseBaseConfig()
+web_config = ConcourseWebConfig()
+worker_config = ConcourseWorkerConfig()
```
Why is this defined here but not being used?

blarghmatey replied:
Just haven't integrated it yet. The plan is to add a switch to determine whether to pull in the config for a worker or a web node at build time and propagate the appropriate config.

blarghmatey committed:
Adding in Pants and starting work to package deploys

In order to be able to support building images via Packer we want to be able to package up the PyInfra deploys so that they can be copied to the target host and executed directly. To make this possible this commit adds the Pants build tool and an initial configuration for packaging the target deployment as a PEX binary.

shaidar commented on Vagrantfile:
```diff
@@ -0,0 +1,80 @@
+  # config.vm.provision "shell", inline: <<-SHELL
+  #   apt-get update
+  #   apt-get install -y apache2
+  # SHELL
```
Any reason/use to keep all those commented lines?

blarghmatey replied:
No particular reason.

blarghmatey committed:
Added a Packer configuration for building the Concourse component.

blarghmatey committed:
Added Copier template for bootstrapping PyInfra components

shaidar commented:
Overall looks good as far as I can tell. I'd imagine once you run the entire workflow you'll notice something here and there, but it's 👍 for me

blarghmatey committed:
Adding init file in component template
